core/amber/src/main/scala/edu/uci/ics/texera/web/resource/ResultDownloadResource.scala

```diff
@@ -0,0 +1,202 @@
+      )
+    }
+    val result: List[ITuple] =
+      sessionResults.map(p => p._2).find(p => !p.isEmpty).get
+
+    var response: ResultDownloadResponse = null
+    request.downloadType match {
+      case "google_sheet" =>
+        response = handleGoogleSheetRequest(request, result)
+      case _ =>
+        response = ResultDownloadResponse(
+          request.downloadType,
+          "",
+          s"Unknown download type: ${request.downloadType}"
+        )
+    }
+
+    // cached the download response in the session
+    if (!response.link.isEmpty) {
+      session.getUserProperties
+        .put(getSessionName(request.downloadType), response)
+    }
+    response
+  }
+
+  private def getSessionName(downloadType: String): String = {
+    s"${SESSION_CACHED_KEY_NAME}_${downloadType}"
+  }
+
+  private def handleGoogleSheetRequest(
+      resultDownloadRequest: ResultDownloadRequest,
+      result: List[ITuple]
+  ): ResultDownloadResponse = {
+    // create google sheet
+    val sheetService: Sheets = GoogleResource.createSheetService()
+    val sheetId: String =
+      createGoogleSheet(sheetService, resultDownloadRequest.workflowName)
+    if (sheetId == null)
+      return ResultDownloadResponse(
+        resultDownloadRequest.downloadType,
+        "",
+        "Fail to create google sheet"
+      )
+
+    // upload the schema
+    val schema: util.List[AnyRef] = getSchema(result)
+    val schemaContent: util.List[util.List[AnyRef]] = Lists.newArrayList()
+    schemaContent.add(schema)
+    val response: AppendValuesResponse =
+      uploadContent(sheetService, sheetId, schemaContent)
+
+    // allow user to access this sheet in the service account
+    val drive: Drive = GoogleResource.createDriveService()
+    val sharePermission: Permission = new Permission()
+      .setType("anyone")
+      .setRole("reader")
+    drive
+      .permissions()
+      .create(sheetId, sharePermission)
+      .execute()
+
+    // upload the content asynchronously to avoid long waiting on the user side.
+    // may change to thread pool
+    new Thread(() => uploadResult(sheetService, sheetId, result)).start()
+
+    // generate success response
+    val link: String = s"https://docs.google.com/spreadsheets/d/$sheetId/edit"
+    val message: String =
+      s"Google sheet created. The results may be still uploading."
+    ResultDownloadResponse(resultDownloadRequest.downloadType, link, message)
+  }
+
+  /**
+    * create the google sheet and return the sheet Id
+    */
+  private def createGoogleSheet(sheetService: Sheets, workflowName: String): String = {
+    val createSheetRequest = new Spreadsheet()
+      .setProperties(new SpreadsheetProperties().setTitle(workflowName))
+    val targetSheet: Spreadsheet = sheetService.spreadsheets
+      .create(createSheetRequest)
+      .setFields("spreadsheetId")
+      .execute
+    targetSheet.getSpreadsheetId
+  }
+
+  private def getSchema(result: List[ITuple]): util.List[AnyRef] = {
```

Comment:
since we implicitly assume that the schema won't change during an execution, can you take in a single Tuple to access its schema instead?


Another thought, as @maazadeeb is trying to decouple `Texera.Schema` from `Texera.Tuple`, is there another way to get a reference of the schema?

Reply:
`take in a single Tuple to access its schema` is exactly what I did inside this function. I just pass the whole list into it to handle the empty input cases. I tried other ways to get the schema but have no clues. If @maazadeeb finish his part then I will be happy to change this code.